src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemorySegment.java

```diff
@@ -321,18 +321,33 @@ default MemorySegment asSlice(long offset) {
     boolean isMapped();
 
     /**
-     * Returns true if this segment overlaps the {@code other} segment.
+     * Returns a new memory segment that is the overlap of the two segments.
      *
-     * If the two segments are not either both backed by native or heap memory, {@code false} is returned.
-     * Otherwise, their offset and {@linkplain #byteSize() size} (in bytes) are used to determine if an overlap occurs.
+     * <p>Two segments S1 and S2 are said to overlap if it is possible to find
+     * at least two slices L1 (from S1) and L2 (from S2) such that L1 and L2 are
+     * backed by the same memory region. As such, it is not possible for an
+     * {@link #isNative() native} segment to overlap with a heap segment; in
+     * this case, or when no overlap occurs, {@code null} is returned.
      *
-     * @param other the segment to be tested for an overlap with this segment.
-     * @return {@code true} if this segment overlaps the other segment.
-     * @throws IllegalStateException if either the scope associated with this segment or the scope associated
-     * with the {@code other} segment have been already closed, or if access occurs from a thread other than the thread
-     * owning either scopes.
+     * @param other the segment to test for an overlap with this segment.
+     * @return a new memory segment, or {@code null} if no overlap occurs.
+     */
+    MemorySegment overlap(MemorySegment other);
+
+    /**
+     * Returns the offset, in bytes, from this segment to the other segment.
+     *
+     * <p>The offset is relative to the {@linkplain #address() base address} of
+     * this segment and will be in the range of 0 (inclusive) up to the
+     * {@linkplain #byteSize() size} (in bytes) of this segment (exclusive). If
+     * the two segments do not {@link #overlap(MemorySegment) overlap}, an
+     * offset can not be retrieved and {@code -1L} is returned.
+     *
+     * @param other the segment to retrieve an offset to.
+     * @return the relative offset, in bytes, from this segment to the other
+     *         segment, or -1L if no overlap occurs.
      */
-    boolean isOverlapping(MemorySegment other);
+    long offsetOf(MemorySegment other);
```

Comment:
I think what I had in mind was slightly different:
* return postive, or negative offsets in case of overlaps (of course in the case of sub-slices, offsets will always be positive)
* throw if asked to compute an offset from heap to non-heap (or vice-versa).